Tribler/Policies/BoostingManager.py

```diff
@@ -0,0 +1,493 @@
+
+        self.session = session
+        assert self.session.get_libtorrent()
+
+        self.torrent_db = self.session.open_dbhandler(NTFY_TORRENTS)
+        self.channelcast_db = self.session.open_dbhandler(NTFY_CHANNELCAST)
+
+        # use provided settings or a default one
+        self.settings = settings or BoostingSettings(session, load_config=True)
+
+        if self.settings.load_config:
+            self._logger.info("Loading config file from session configuration")
+            self.load_config()
+
+        if not os.path.exists(self.settings.credit_mining_path):
+            os.makedirs(self.settings.credit_mining_path)
+
+        self.session.lm.ltmgr.get_session().set_settings(
+            {'share_mode_target': self.settings.share_mode_target})
+
+        self.session.add_observer(self.on_torrent_notify, NTFY_TORRENTS, [NTFY_UPDATE])
+
+        self.register_task("CreditMining_select", LoopingCall(self._select_torrent),
+                           self.settings.initial_swarm_interval, interval=self.settings.swarm_interval)
+
+        self.register_task("CreditMining_scrape", LoopingCall(self.scrape_trackers),
+                           self.settings.initial_tracker_interval, interval=self.settings.tracker_interval)
+
+        self.register_task("CreditMining_log", LoopingCall(self.log_statistics),
+                           self.settings.initial_logging_interval, interval=self.settings.logging_interval)
+
+    @staticmethod
+    def get_instance(*args, **kw):
+        """
+        get single instance of Boostingmanager
+        """
+        if BoostingManager.__single is None:
+            BoostingManager(*args, **kw)
+        return BoostingManager.__single
+
+    @staticmethod
+    def del_instance():
+        """
+        resetting, then deleting single instance
+        """
+        BoostingManager.__single = None
+
+    def shutdown(self):
+        """
+        save configuration before stopping stuffs
+        """
+        self.save_config()
+        self._logger.info("Shutting down boostingmanager")
+
+        for sourcekey in self.boosting_sources.keys():
+            self.remove_source(sourcekey)
+
+        self.cancel_all_pending_tasks()
+
+        #remove credit mining data in not persistent mode
+        shutil.rmtree(self.settings.credit_mining_path, ignore_errors=True)
+
+    def get_source_object(self, sourcekey):
+        """
+        Get the actual object of the source string/key
+        """
+        # first, assume it's already source-friendly format
+        ret = self.boosting_sources.get(sourcekey, None)
+
+        # we can't find source, lets assume it's string now
+        if not ret:
+            try:  # is it channel?
+                int(sourcekey, 16)
+                ret = self.boosting_sources.get(unhexlify(sourcekey), None)
+            except ValueError:
+                # it's not, so we can't find it
+                self._logger.debug("Can't find source object in type %s", type(sourcekey))
```

Comment:
If you can't find the source it means that something went wrong doesn't it? It should not happen.